editor/modes/visual-editor/block.js

```diff
@@ -210,7 +251,13 @@ class VisualEditorBlock extends Component {
 
 		// Deselect on escape.
 		if ( ESCAPE === keyCode ) {
-			onDeselect();
+			if ( isToolbar( target ) ) {
```

Comment:
I rejected option 1 as being inefficient.

Another option is for the toolbars to fire custom events instead of a prop callback. The problem with prop callbacks are that the toolbars are not the direct child of the block.